File: RSIStrategy.java
import java.math.BigDecimal;
import java.util.List;

/**
 * RSI (Relative Strength Index) mean-reversion strategy.
 *
 * <ul>
 *   <li><b>BUY:</b> RSI crosses above the oversold threshold (recovery from oversold)</li>
 *   <li><b>SELL:</b> RSI crosses below the overbought threshold (retreat from overbought)</li>
 * </ul>
 *
 * <p>Common configurations: period=14, oversold=30, overbought=70.</p>
 */
public final class RSIStrategy implements TradingStrategy {

    private final int period;
    private final BigDecimal oversoldThreshold;
    private final BigDecimal overboughtThreshold;

    public RSIStrategy(int period, double oversold, double overbought) {
        if (period <= 0) {
            throw new IllegalArgumentException("RSI period must be positive: " + period);
        }
        if (oversold < 0 || oversold > 100 || overbought < 0 || overbought > 100) {
            throw new IllegalArgumentException(
                    "Thresholds must be in [0, 100]: oversold=%s, overbought=%s"
                            .formatted(oversold, overbought));
        }
        if (oversold >= overbought) {
            throw new IllegalArgumentException(
                    "Oversold (%s) must be less than overbought (%s)".formatted(oversold, overbought));
        }

        this.period = period;
        this.oversoldThreshold = BigDecimal.valueOf(oversold);
        this.overboughtThreshold = BigDecimal.valueOf(overbought);
    }

    @Override
    public String getName() {
        return "RSI (%d) [%s/%s]".formatted(period, oversoldThreshold, overboughtThreshold);
    }

    @Override
    public Signal evaluate(List<PriceData> data, int currentIndex) {
        validateEvaluationInputs(data, currentIndex);

        if (!isWarmedUp(currentIndex)) {
            return Signal.HOLD;
        }

        var currentRSI = TechnicalIndicators.rsi(data, currentIndex, period);
        var prevRSI = TechnicalIndicators.rsi(data, currentIndex - 1, period);

        // Buy: RSI crosses above oversold threshold (recovery)
        if (currentRSI.compareTo(oversoldThreshold) > 0
                && prevRSI.compareTo(oversoldThreshold) <= 0) {
            return Signal.BUY;
        }

        // Sell: RSI crosses below overbought threshold (retreat)
        if (currentRSI.compareTo(overboughtThreshold) < 0
                && prevRSI.compareTo(overboughtThreshold) >= 0) {
            return Signal.SELL;
        }

        return Signal.HOLD;
    }

    @Override
    public int getWarmupPeriod() {
        return period + 1;
    }

    public int getPeriod()                    { return period; }
    public BigDecimal getOversoldThreshold()  { return oversoldThreshold; }
    public BigDecimal getOverboughtThreshold() { return overboughtThreshold; }

    @Override
    public String toString() {
        return getName();
    }
}

File: ios/StockAnalyzer/StockAnalyzer/Strategies/RSIStrategy.swift
import Foundation

struct RSIStrategy30_70: TradingStrategy {
    let name = "RSI (30/70)"
    let description = "Mean reversion: Buys when RSI drops below 30, sells when RSI rises above 70"
    let warmupPeriod = 15

    func evaluate(data: [PriceData], index: Int, indicators: StrategyIndicators) -> TradeAction {
        guard let rsi = indicators.rsi14[index] else {
            return .hold
        }

        if rsi < 30 {
            return .buy
        }

        if rsi > 70 {
            return .sell
        }

        return .hold
    }
}

struct RSIStrategy25_75: TradingStrategy {
    let name = "RSI (25/75)"
    let description = "Aggressive RSI: Buys when RSI drops below 25, sells when RSI rises above 75"
    let warmupPeriod = 15

    func evaluate(data: [PriceData], index: Int, indicators: StrategyIndicators) -> TradeAction {
        guard let rsi = indicators.rsi14[index] else {
            return .hold
        }

        if rsi < 25 {
            return .buy
        }

        if rsi > 75 {
            return .sell
        }

        return .hold
    }
}
